File: 2000 CCC/s5.cpp
#include <iostream>
#include <vector>
#include <iomanip> // for std::fixed and std::setprecision

int main() {
    std::vector<double> x(100);
    std::vector<double> y(100);
    std::vector<bool> out(100, false);  // list of "out" sheep

    int n;
    std::cin >> n;
    for (int i = 0; i < n; i++) {
        std::cin >> x[i] >> y[i];
    }

    for (int i = 0; i < n; i++) {
        double left = 0, right = 1000;
        for (int j = 0; j < n; j++) {
            if (!out[i] && !out[j] && i != j) {
                double xm = (x[i] + x[j]) / 2;
                double ym = (y[i] + y[j]) / 2;
                double s = (x[i] - x[j]) / (y[j] - y[i]);
                if (s == 0) {
                    if (y[i] < y[j])
                        out[j] = true;
                    else
                        out[i] = true;
                }
                else {
                    double p = -ym / s + xm;
                    if (x[j] < x[i])
                        left = std::max(p, left);
                    else
                        right = std::min(p, right);
                }
            }
        }
        if (left >= right)
            out[i] = true;
    }

    for (int j = 0; j < n; j++) {
        if (!out[j]) {
            std::cout << "The sheep at (" << std::fixed << std::setprecision(2) << x[j] << ", " << y[j] << ") might be eaten." << std::endl;
        }
    }

    return 0;
}

File: 2000 CCC/s5.py
# sheep and coyotes

sheep = []
eaten = []

for _ in range(int(input())):
    sheep.append([float(input()), float(input())])

for i in range(len(sheep)):
    left, right = 0, 1000
    
    for j in range(len(sheep)):
        if i not in eaten and j not in eaten and i != j and sheep[j][1] != sheep[i][1]:
            x = (sheep[i][0] + sheep[j][0]) / 2
            y = (sheep[i][1] + sheep[j][1]) / 2
            s = (sheep[i][0] - sheep[j][0]) / (sheep[j][1] - sheep[i][1])
            if s == 0:
                if sheep[i][1] < sheep[j][1]:
                    eaten.append(j)
                else:
                    eaten.append(i)
            else:
                p = -y / s + x
                if sheep[j][0] < sheep[i][0]:
                    left = max(p, left)
                else:
                    right = min(p, right)

    if left >= right:
        eaten.append(i)

for j in range(len(sheep)):
    if j not in eaten:
        print('The sheep at (', format(sheep[j][0],'.2f'), ', ', format(sheep[j][1],'.2f'), ') might be eaten.', sep='')
